File: src/main/java/cc/mirukuneko/realtrainmodrenewed/entity/formation/FormationManager.java
package cc.mirukuneko.realtrainmodrenewed.entity.formation;

import cc.mirukuneko.realtrainmodrenewed.entity.TrainEntity;
import java.util.HashMap;
import java.util.Map;

public final class FormationManager {
    private static FormationManager INSTANCE;
    private final Map<Long, Formation> formations = new HashMap<>();

    private FormationManager() {}

    public static FormationManager getInstance() {
        if (INSTANCE == null) INSTANCE = new FormationManager();
        return INSTANCE;
    }

    public static void reset() { INSTANCE = null; }

    public Formation getFormation(long id) { return formations.get(id); }
    public void register(long id, Formation f) { formations.put(id, f); }
    public void remove(long id) { formations.remove(id); }

    public long getNewId() {
        long id = System.currentTimeMillis();
        // ensure uniqueness
        while (formations.containsKey(id)) id++;
        return id;
    }

    public Formation createNewFormation(TrainEntity train) {
        long id = getNewId();
        Formation f = new Formation(id, 1);
        f.entries[0] = new FormationEntry(train, 0, 0);
        train.setFormation(f);
        return f;
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/entity/formation/FormationManager.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.entity.formation

import cc.mirukuneko.realtrainmodrenewed.entity.TrainEntity

class FormationManager private constructor() {
    private val formations: MutableMap<String, Formation> = HashMap()

    fun getFormation(id: String): Formation? = formations[id]

    fun register(id: String, formation: Formation) {
        formations[id] = formation
    }

    fun remove(id: String) {
        formations.remove(id)
    }

    fun getNewId(): Long {
        var id = System.currentTimeMillis()
        while (formations.containsKey(id.toString())) {
            id++
        }
        return id
    }

    fun createNewFormation(train: TrainEntity?): Formation? {
        if (train == null) return null
        val id = getNewId().toString()
        val formation = Formation(id, 1)
        formation.entries[0] = FormationEntry(train, 0, 0)
        return formation
    }

    companion object {
        @JvmField
        val instance: FormationManager = FormationManager()

        @JvmStatic
        fun getInstance(): FormationManager = instance
    }
}
